debezium-core/src/main/java/io/debezium/heartbeat/DatabaseHeartbeatImpl.java

```diff
@@ -47,8 +49,16 @@ public void forcedBeat(Map<String, ?> partition, Map<String, ?> offset, Blocking
         try {
             jdbcConnection.execute(heartBeatActionQuery);
         }
-        catch (Exception e) {
-            LOGGER.error("Could not execute heartbeat action", e);
+        catch (SQLException e) {
+            String sqlErrorId = e.getSQLState();
+            switch (sqlErrorId) {
+                case "57P01":  // Postgres error admin_shutdown, see https://www.postgresql.org/docs/9.4/errcodes-appendix.html
```

Comment:
Checking Postgres error codes inside `debezium-core` might smell, but it solves the issue. WDYT?
I could add a constructor parameter and inject a List<String> of heartbeat errors which should trigger an exception but I don't know if you think it's worth?

Reply:
This is definitely not a ood solution. It woud be valid to change Heartbeat contract in the way that for example you'd pass a code that would do the evaluation you do right now. In the similar way as is done in `io.debezium.pipeline.ErrorHandler`.

Also do you really want to throw the exception? It sounds like a real hack here. In fact it should be a callback that will be passed from the connector. When this happens the callback will change a flag monitored in the main connector loop or ideall in `io.debezium.pipeline.ChangeEventSourceCoordinator.ChangeEventSourceContextImpl` or call connector stop. This will trigger graceful connector shutdown

Reply:
Thank you @jpechane. I applied most parts of you feedback, I guess.
I will reply on the exception vs graceful stop in the refactored code.